File: Selector.cs
public class Selector<T>
{
	private Inventor.Application _inventor;

	public Selector(Inventor.Application ThisApplication)
	{
		_inventor = ThisApplication;

		InteractProperties = _inventor.CommandManager.CreateInteractionEvents();
		InteractProperties.InteractionDisabled = false;
		InteractProperties.OnTerminate += oInteractEvents_OnTerminate;

		SelectProperties = InteractProperties.SelectEvents;
		SelectProperties.WindowSelectEnabled = false;
		SelectProperties.OnSelect += oSelectEvents_OnSelect;
		SelectProperties.OnPreSelect += oSelectEvents_OnPreSelect;

		SelectionMethod = DefaultSelector;
	}

	public void Pick()
	{
		SelectionStatus = SelectionStatusEnum.StillSelecting;
		InteractProperties.Start();
		while (SelectionStatus == SelectionStatusEnum.StillSelecting)
			_inventor.UserInterfaceManager.DoEvents();
		InteractProperties.Stop();
		_inventor.CommandManager.StopActiveCommand();
	}

	public InteractionEvents InteractProperties { get; set; }
	public SelectEvents SelectProperties { get; set; }
	public IEnumerable<T> SelectedEntities { get; set; } = null;
	public Inventor.Point ModelPosition { get; set; } = null/* TODO Change to default(_) if this is not a reference type */;
	public Point2d ViewPosition { get; set; } = null/* TODO Change to default(_) if this is not a reference type */;
	public Func<object, bool> SelectionMethod { get; set; }
	public SelectionStatusEnum SelectionStatus { get; set; } = SelectionStatusEnum.NotStarted;

	private bool DefaultSelector(object PreSelectEntity)
	{
		return (PreSelectEntity is T);
	}

	private void oSelectEvents_OnPreSelect(ref object PreSelectEntity, out bool DoHighlight, ref ObjectCollection MorePreSelectEntities, SelectionDeviceEnum SelectionDevice, Inventor.Point ModelPosition, Point2d ViewPosition, Inventor.View View)
	{
		DoHighlight = SelectionMethod(PreSelectEntity);
	}

	private void oInteractEvents_OnTerminate()
	{
		if ((SelectionStatus != SelectionStatusEnum.SometingIsSelected))
			SelectionStatus = SelectionStatusEnum.Canceled;
	}

	private void oSelectEvents_OnSelect(ObjectsEnumerator JustSelectedEntities, SelectionDeviceEnum SelectionDevice, Inventor.Point ModelPosition, Point2d ViewPosition, Inventor.View View)
	{
		List<T> selectedObjects = new List<T>();
		foreach (object item in JustSelectedEntities)
		{
			if ((SelectionMethod(item)))
			{
				selectedObjects.Add((T)item);
				this.ModelPosition = ModelPosition;
			}
		}

		this.SelectedEntities = selectedObjects;
		this.ModelPosition = ModelPosition;
		this.ViewPosition = ViewPosition;

		SelectionStatus = SelectionStatusEnum.SometingIsSelected;
	}
}

public enum SelectionStatusEnum
{
	NotStarted,
	StillSelecting,
	Canceled,
	SometingIsSelected
}

File: Selector.vb
Public Class Selector(Of T)

    Private _inventor As Inventor.Application

    Public Sub New(ThisApplication As Inventor.Application)
        _inventor = ThisApplication

        InteractProperties = _inventor.CommandManager.CreateInteractionEvents
        InteractProperties.InteractionDisabled = False
        AddHandler InteractProperties.OnTerminate, AddressOf oInteractEvents_OnTerminate

        SelectProperties = InteractProperties.SelectEvents
        SelectProperties.WindowSelectEnabled = False
        AddHandler SelectProperties.OnSelect, AddressOf oSelectEvents_OnSelect
        AddHandler SelectProperties.OnPreSelect, AddressOf oSelectEvents_OnPreSelect

        SelectionMethod = AddressOf DefaultSelector
    End Sub

    Public Sub Pick()
        SelectionStatus = SelectionStatusEnum.StillSelecting
        InteractProperties.Start()
        Do While SelectionStatus = SelectionStatusEnum.StillSelecting
            _inventor.UserInterfaceManager.DoEvents()
        Loop
        InteractProperties.Stop()
        _inventor.CommandManager.StopActiveCommand()
    End Sub

    Public Property InteractProperties As InteractionEvents
    Public Property SelectProperties As SelectEvents

    Public Property SelectedEntities As IEnumerable(Of T) = Nothing
    Public Property ModelPosition As Point = Nothing
    Public Property ViewPosition As Point2d = Nothing

    Public Property SelectionMethod As Func(Of Object, Boolean)
    Public Property SelectionStatus As SelectionStatusEnum = SelectionStatusEnum.NotStarted

    Private Function DefaultSelector(PreSelectEntity As Object) As Boolean
        Return (TypeOf PreSelectEntity Is T)
    End Function

    Private Sub oSelectEvents_OnPreSelect(
            ByRef PreSelectEntity As Object,
            ByRef DoHighlight As Boolean,
            ByRef MorePreSelectEntities As ObjectCollection,
            SelectionDevice As SelectionDeviceEnum,
            ModelPosition As Point,
            ViewPosition As Point2d,
            View As Inventor.View)

        DoHighlight = SelectionMethod(PreSelectEntity)
    End Sub

    Private Sub oInteractEvents_OnTerminate()
        If (SelectionStatus <> SelectionStatusEnum.SometingIsSelected) Then
            SelectionStatus = SelectionStatusEnum.Canceled
        End If
    End Sub

    Private Sub oSelectEvents_OnSelect(
            ByVal JustSelectedEntities As ObjectsEnumerator,
            ByVal SelectionDevice As SelectionDeviceEnum,
            ByVal ModelPosition As Point,
            ByVal ViewPosition As Point2d,
            ByVal View As Inventor.View)

        Dim selectedObjects As New List(Of T)
        For Each item As Object In JustSelectedEntities
            If (SelectionMethod(item)) Then
                selectedObjects.Add(item)
                Me.ModelPosition = ModelPosition
            End If
        Next

        Me.SelectedEntities = selectedObjects
        Me.ModelPosition = ModelPosition
        Me.ViewPosition = ViewPosition

        SelectionStatus = SelectionStatusEnum.SometingIsSelected
    End Sub
	
    ' Copyright 2024
    ' 
    ' This code was written by Jelte de Jong, and published on www.hjalte.nl
    '
    ' Permission Is hereby granted, free of charge, to any person obtaining a copy of this 
    ' software And associated documentation files (the "Software"), to deal in the Software 
    ' without restriction, including without limitation the rights to use, copy, modify, merge, 
    ' publish, distribute, sublicense, And/Or sell copies of the Software, And to permit persons 
    ' to whom the Software Is furnished to do so, subject to the following conditions:
    '
    ' The above copyright notice And this permission notice shall be included In all copies Or
    ' substantial portions Of the Software.
    ' 
    ' THE SOFTWARE Is PROVIDED "AS IS", WITHOUT WARRANTY Of ANY KIND, EXPRESS Or IMPLIED, 
    ' INCLUDING BUT Not LIMITED To THE WARRANTIES Of MERCHANTABILITY, FITNESS For A PARTICULAR 
    ' PURPOSE And NONINFRINGEMENT. In NO Event SHALL THE AUTHORS Or COPYRIGHT HOLDERS BE LIABLE 
    ' For ANY CLAIM, DAMAGES Or OTHER LIABILITY, WHETHER In AN ACTION Of CONTRACT, TORT Or 
    ' OTHERWISE, ARISING FROM, OUT Of Or In CONNECTION With THE SOFTWARE Or THE USE Or OTHER 
    ' DEALINGS In THE SOFTWARE.
End Class
Public Enum SelectionStatusEnum : NotStarted : StillSelecting : Canceled : SometingIsSelected : End Enum